### PR TITLE
[Routing] fix conflict with param named class in attribute

### LINKED ISSUE #40225: conflict with AnnotationFileLoader and PHP8 attributes
**Symfony version(s) affected**: 5.2.x

**PHP version**: 8.x

**Description**  
When an attribute has a parameter named class before the class declaration, the script tries to load the content of that parameter.

```
...
#[ParamConverter(
    'availability',
    class: Table\Availability::class,
    options: ['id' => 'ida']
)]
...
class EditController
{
...
```

**How to reproduce**  
See issue https://github.com/sensiolabs/SensioFrameworkExtraBundle/issues/713 for more detail

**Possible Solution**  
See here https://github.com/sensiolabs/SensioFrameworkExtraBundle/issues/713#issuecomment-780669307

### PR DESCRIPTION
**Pull request by nlhommet**
| Q             | A
| ------------- | ---
| Branch        | 4.4 
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Fix #40225
| License       | MIT
| Doc PR        | -

Fix conflict with AnnotationFileLoader and class PHP8 Attribute with param named "class"

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Can you please rebase+retarget on 5.2 if that's really the target?

**Comment by nlhommet** [comment]
Sorry it's my first PR, can you tell me how to correct my PR to be on 5.2?

**Comment by nicolas-grekas** [comment]
I force-pushed on your fork.
FYI, I used `git rebase -i` to do the rebase on 5.2.
Do you think you could add a test case, please?
(don't miss fetching your remote branch first of course)

**Comment by derrabus** [comment]
Are we sure that 4.4 is not affected?

**Comment by nicolas-grekas** [comment]
(now targetting 4.4)

**Comment by nlhommet** [comment]
Tx for the help.
This impacts all branches compatible with PHP8 Attributes.
I set up tests today.

**Comment by nlhommet** [comment]
I need to understand why it does not work in PHP7.1 on multiline Attributes

File: src/Symfony/Component/Routing/Tests/Fixtures/Attributes/FooAttributes.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Tests/Fixtures/Attributes/FooAttributes.php]: please remove them all, the codebase doesn't use "strict mode".

**Comment by nlhommet** [comment]
I think there are tests that can only be done in PHP8.
Because an Attributes multiline like this 
```
#[FooAttributes(
    class: User::class,
    foo: [
        'bar' => ['foo','bar'],
        'foo'
    ]
)]
```
is not recognized as a valid instruction in PHP7.X.

With `#[FooAttributes(class: User::class,foo: ['bar' => ['foo','bar'],'foo'])]`, the Attribute is recognized as comment in PHP7.X and therefore poses no problem.

**Comment by nicolas-grekas** [comment]
> I think there are tests that can only be done in PHP8.

correct. You can add the `@requires PHP 8` annotation on top of those tests.

**Comment by derrabus** [comment]
Thank you for working on this. However, I'm not convinced that bracket counting is the way we should fix this. We already have a piece of code in that file that handles incorrect `T_CLASS` tokens and I think we can extend that logic.

```patch
diff --git a/src/Symfony/Component/Routing/Loader/AnnotationFileLoader.php b/src/Symfony/Component/Routing/Loader/AnnotationFileLoader.php
--- a/src/Symfony/Component/Routing/Loader/AnnotationFileLoader.php	(revision eeed8c8e847b81ea5d16685395e85a1df9077449)
+++ b/src/Symfony/Component/Routing/Loader/AnnotationFileLoader.php	(date 1613917175487)
@@ -100,20 +100,8 @@
             $nsTokens[T_NAME_QUALIFIED] = true;
         }
 
-        $openBrackets = 0;
         for ($i = 0; isset($tokens[$i]); ++$i) {
             $token = $tokens[$i];
-            if (80000 <= \PHP_VERSION_ID) {
-                if (\T_ATTRIBUTE === ($token[0] ?? 0) || ($openBrackets > 0 && '[' === $token)) {
-                    ++$openBrackets;
-                }
-                if ($openBrackets > 0 && ']' === $token) {
-                    --$openBrackets;
-                }
-                if ($openBrackets > 0) {
-                    continue;
-                }
-            }
 
             if (!isset($token[1])) {
                 continue;
@@ -146,6 +134,15 @@
                         break;
                     }
                 }
+                for ($j = $i + 1; isset($tokens[$j]); ++$j) {
+                    if (':' === $tokens[$j]) {
+                        $skipClassToken = true;
+                        break;
+                    }
+                    if (!\in_array($tokens[$j][0], [\T_WHITESPACE, \T_DOC_COMMENT, \T_COMMENT])) {
+                        break;
+                    }
+                }
 
                 if (!$skipClassToken) {
                     $class = true;
```

**Comment by nlhommet** [comment]
With this test I check if the `T_CLASS` token is after a comma or a parenthesis
This covers these 2 scenarios :
```
#[FooAttributes(
    foo: [
        'bar' => ['foo','bar'],
        'foo'
    ],
    class: User::class
)]
```
```
#[FooAttributes(
    class: User::class,
    foo: [
        'bar' => ['foo','bar'],
        'foo'
    ]
)]
```
Normally it shouldn't have side effects

**Review by derrabus — APPROVED** [review]
Please make fabbot happy. Apart from that, I think we're good. Thank you for working on this issue.

**Comment by nicolas-grekas** [comment]
Thank you @nlhommet.

**Pull request merged**